hack/libvirt-ca/main.go

```diff
@@ -0,0 +1,206 @@
+package main
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+
+	"github.com/apparentlymart/go-cidr/cidr"
+	"github.com/openshift/installer/pkg/asset/tls"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+const (
+	orgUnit          = "openshift"
+	caCommonName     = "libvirt"
+	serverCommonName = "libvirt"
+	clientCommonName = "openshift-cluster-api"
+	defaultNetwork   = "192.168.124.0/24"
+	defaultOutDir    = "."
+
+	// Libvirt is picky about the naming here:
+	// https://libvirt.org/remote.html
+	caCertFile     = "cacert.pem"
+	caKeyFile      = "cakey.pem"
+	serverCertFile = "servercert.pem"
+	serverKeyFile  = "serverkey.pem"
+	clientCertFile = "clientcert.pem"
+	clientKeyFile  = "clientkey.pem"
+)
+
+var (
+	network = kingpin.Flag("network", "Cluster network CIDR.").
+		Short('n').Default(defaultNetwork).String()
+
+	outDir = kingpin.Flag("out", "Output directory.").
+		Short('o').Default(defaultOutDir).ExistingDir()
+)
+
+type certificate struct {
+	key  *rsa.PrivateKey
+	cert *x509.Certificate
+}
+
+func (c *certificate) WritePEMs(certPath, keyPath string) error {
+	if err := ioutil.WriteFile(keyPath, []byte(tls.PrivateKeyToPem(c.key)), 0600); err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(certPath, []byte(tls.CertToPem(c.cert)), 0644); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getGateway(network string) (net.IP, error) {
+	_, ipNet, err := net.ParseCIDR(network)
+	if err != nil {
+		return nil, err
+	}
+
+	gateway, err := cidr.Host(ipNet, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	return gateway, nil
+}
+
+func generateCA() (*certificate, error) {
+	var ca certificate
+	var err error
+
+	cfg := &tls.CertCfg{
+		Subject: pkix.Name{
+			CommonName:         caCommonName,
+			OrganizationalUnit: []string{orgUnit},
+		},
+		KeyUsages: x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		Validity:  tls.ValidityTenYears,
+		IsCA:      true,
+	}
+
+	ca.key, ca.cert, err = tls.GenerateRootCertKey(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ca, err
+}
+
+func generateServerCert(dnsNames []string, ips []net.IP, ca *certificate) (*certificate, error) {
+	var server certificate
+	var err error
+
+	cfg := &tls.CertCfg{
+		KeyUsages:    x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
```

Comment:
Does libvirt explicitly require a client auth EKU on the server cert?

Reply:
Oops. No. Fixed.